java-type-checker/java_type_checker/types.py

```diff
@@ -133,25 +140,49 @@ def method_named(self, name):
                     pass
             raise NoSuchJavaMethod("{0} has no method named {1}".format(self.name, name))
 
+    def is_subtype_of(self, other):
+        if other == self:
+            return True
+        for supertype in self.direct_supertypes:
+            if supertype.is_subtype_of(other):
+                return True
+        return False
+
 
 class JavaVoidType(JavaType):
     """The Java type `void`.
 
     It is never legal to use the result of a method returning void inside a larger expression.
     Void is therefore subtype only of itself, and not any other type.
     """
+
     def __init__(self):
         super().__init__("void")
 
+    def is_subtype_of(self, other):
+        if other == self:
+            return True
+        return False
+
 
 class JavaNullType(JavaType):
     """The type of the value `null` in Java.
 
     Null acts as though it is a subtype of all object types. However, it raises an exception for any
     attempt to look up a method.
     """
+
     def __init__(self):
         super().__init__("null")
+        self.is_object_type = True
+
+    def is_subtype_of(self, other):
+        if other.is_object_type:
+            return True
+        return False
+
+    def method_named(self, method_name):
+        raise NoSuchJavaMethod("Cannot invoke method " + method_name + "() on null")
```

Comment:
I used "null", while you used `self.name`. `self.name` should always be null, so I don't think there's a difference. Also, the is_subtype_of(self, other) method should have done `return other.is_object_type`.

Reply:
Always beware of repeating the same constant in multiple places in the code, even if you think the redundancy won’t hurt anything. (What if the language decides it’s called `nil` instead?)